instantly/params.go

```diff
@@ -0,0 +1,65 @@
+package instantly
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/amp-labs/connectors/common"
+	"github.com/amp-labs/connectors/common/paramsbuilder"
+	"github.com/amp-labs/connectors/providers"
+)
+
+const (
+	// DefaultPageSize is number of elements per page.
+	DefaultPageSize = 100
+)
+
+// Option is a function which mutates the connector configuration.
+type Option = func(params *parameters)
+
+// parameters surface options by delegation.
+type parameters struct {
+	paramsbuilder.Client
+	// Error is set when any With<Method> fails, used for parameters validation.
+	setupError error
+}
+
+func (p parameters) ValidateParams() error {
+	if p.setupError != nil {
+		return p.setupError
+	}
+
+	return errors.Join(
+		p.Client.ValidateParams(),
+	)
+}
+
+func WithClient(ctx context.Context, client *http.Client,
```

Comment:
This method so far looked identical, but there are more non-oauth2 connectors added which would change the arguments and the impl of the function.

We are using `paramsbuilder.Client` as a delegate. So each connector then selects one of `WithOauthClient` / `WithBasicClient` / `WithApiKeyHeaderClient` / `WithApiKeyQueryParamClient`.

Note: `paramsbuilder` has a dependency on `common` package. In case we need to move such methods somewhere it will be new package to avoid circular dependency. (I tried some refactoring while working on this PR and thats what I had run into).

I used delegation to have some consistency in the way parameters are defined. Since we are using Options pattern the repitition will be there.

Reply:
I looked into the code, I think we can define those functions once in. Looking at how much lines we are replicating for each connector, I think this design is wrong. We always need to think in scale. We will end up having 1000+ connectors. which means this will end up being hundreds of thousands of lines. Which is simply not scalable. I am not sure how we came up this design from the beginning, but if this is current way, we should stop and redesign this. 

We shouldn't declare `type Option = func(*parameters)` type for every connector. 

Option should be in a separate package so that those can be imported and reused, not redefined. 

Let's hold off on this until I can think through and redesign this.

Reply:
First of all by design we require setting up connectors using  `With<MethodName>`. This means every connector must define this method. Also IDE completion makes it useful when say instantly.With... and it autocompletes potential arguments.

`Option`s golang alias is necessary for generic constructor. It provides typed parameter. Every connector must define its set of parameters. So there is nothing that can be refactored here. In other words type `Option = func(*connectorName.parameters)` is not common or shared in any way, only the visual is.

And about methods that live under `parse.go`. This file is responsible to understand response format of a provider. This again cannot be known universally and every connector stores array of items under different field name, everybody does pagination in their unique way. Maybe as more patterns become apparent we could add utility functions to shorten it.

The patterns that could evolve could be more configuration/data driven. But thats the information that makes the identity of that connector. I agree that adding connectors can be confusing especially to outsiders. But slowly these things are polished. I think what we did with `errors.go` can be done here (config driven).